### ISSUE TITLE
Add support for median function in SQL

### ISSUE THREAD

**Issue by newskooler** [questdb/questdb#2830]
### Is your feature request related to a problem?

It's very easy to compute some aggregation functions, such as `avg` or `max`, but other are harder, such as `median`. Having native support in QuestDB would help a lot.

### Describe the solution you'd like.

Ideally it would be called `median` and would work in the same way as the other aggregation functions: https://questdb.io/docs/reference/function/aggregation/

### Describe alternatives you've considered.

I can "manually" do something like: sort the data in descending order and select the top 50 percent of the data and then select the last value.

**Comment by puzpuzpuz** [comment]
As a more Postgres-compatible option, we could add support for the `percentile_cont` [function](https://www.postgresql.org/docs/9.4/functions-aggregate.html). Then calculating the median would be as simple as the following:
```sql
SELECT percentile_cont(0.5) WITHIN GROUP (ORDER BY sort_expression) as median FROM my_tab;
```

**Comment by newskooler** [comment]
Often I'd have to make a bunch of aggregations to data.
For example:
```
SELECT max(price) price_high, avg(price) price_mean, median(price) price_median FROM table
```
What you suggest would not let me do that, right? I will have to do another select on top of an existing select? Possibly even more complicated than that. Is that right?

**Comment by puzpuzpuz** [comment]
With `percentile_cont` your query should look like the following:
```sql
SELECT max(price) price_high, avg(price) price_mean, percentile_cont(0.5) WITHIN GROUP (ORDER BY price) price_median FROM table;
```

Am I missing something?

**Comment by newskooler** [comment]
If it works like this, then yes - that would be fine! : )

**Comment by newskooler** [comment]
Though I would still suggest the following:
1. Change `avg` -> `mean` because `avg` is ambiguous.
2. Have a higher level method of `percentile_cont(0.5) WITHIN GROUP (ORDER BY price)` -> `median`.

Both of these should improve easy of use.

**Comment by puzpuzpuz** [comment]
> 1. Change `avg` -> `mean` because `avg` is ambiguous.

We favor Postgres SQL dialect which calls it `avg`.

**Comment by marregui** [comment]
low hunging t-shirt

**Comment by kevinmingtarja** [comment]
Hi, I'd like to contribute to this issue!

**Comment by puzpuzpuz** [comment]
@kevinmingtarja that's great! I don't think there is anyone working on it.

**Comment by marregui** [comment]
@kevinmingtarja that is fantastic! thank you. The issue has been assigned to you, now you lead. Good luck! if you run into walls please do not hesitate to interact with us as you need.

**Comment by kevinmingtarja** [comment]
Thank you! 

This is my first time contributing, so I'll take some time first to read the codebase. Do you think `core/src/main/java/io/questdb/griffin/engine/functions/groupby/*` is a good place to start for this issue? I see that that's where all the aggregate functions are defined.

**Comment by puzpuzpuz** [comment]
@kevinmingtarja this feature may need changes from #3634 unless you use another way to calculate the median value. Also, please check this thread for some pg-specific alternatives to the median function.

**Comment by kevinmingtarja** [comment]
@puzpuzpuz yeah I plan to implement this api: `percentile_cont(fraction) WITHIN GROUP (ORDER BY sort_expression)` that you pointed out in this thread earlier, so that it's consistent with the postgres api. As for the implementation, I think we have two options:

1. I can build on top of #3634, we can use the `Histogram` added in that PR to calculate the percentile. But it seems like most of the code comes from an external library, so I'm not sure if that is desirable or not
2. If we don't want to build on top of that PR, I can implement the functionality myself. A straightforward solution would be to either do Quickselect or sort the full data set, then finding the percentile is trivial.

**Comment by kevinmingtarja** [comment]
I've taken a look at what postgres and some others have done, and I'll summarize it here. The way I understand it, we can do either **exact** or **approximate** percentiles. 

For exact percentiles, there's no choice but to sort the whole data set. The way postgres [1, 2] does it is it first checks if all the tuples fit within the allowed amount of memory (`work_mem`), if yes, then just do quicksort. Else, they sort using temporary files in disk and external sort (k-way merge, similar to merge sort). It also can’t be sped up by indexes.

Some databases choose to implement approximate percentiles instead, which is quicker and consumes less memory, and they argued that the trade-offs are worthwhile. Timescale [3] and Citus Data [4] has a detailed blog post on this. The two well known algorithms for calculating approx percentiles are HDR Histograms (which is the one in PR #3634), and t-digest.

With this in mind, I'm proposing that we implement both exact and approximate percentiles. We can name the approximate one something like `approx_percentile_cont`. The other PR will help to implement approximate percentiles, so I can do that first. Then we can revisit the exact percentiles after.

---
References:
[1] https://github.com/postgres/postgres/blob/master/src/backend/utils/adt/orderedsetaggs.c#L528-L609
[2] https://github.com/postgres/postgres/blob/master/src/backend/utils/sort/tuplesort.c#L1382C16
[3] https://timescale.com/blog/how-percentile-approximation-works-and-why-its-more-useful-than-averages/
[4] https://techcommunity.microsoft.com/t5/azure-database-for-postgresql/diary-of-an-engineer-delivering-45x-faster-percentiles-using/ba-p/1685102

**Comment by newskooler** [comment]
> > 1. Change `avg` -> `mean` because `avg` is ambiguous.
> 
> We favor Postgres SQL dialect which calls it `avg`.

That makes sense. I do highlight that people who compute mathematical expressions will find it ambiguous and favor precision IMO.

**Comment by puzpuzpuz** [comment]
Heads-up: our next release will feature [`approx_percentile`](3900) function which can be used to calculate an approximate median:
```sql
SELECT approx_percentile(price, 0.5) as median_price FROM prices;
```

**Comment by kevinmingtarja** [comment]
Hi @puzpuzpuz on a side note, should I make a change to the docs too to document this new function?

**Comment by puzpuzpuz** [comment]
@kevinmingtarja thanks for willing to contribute. Unfortunately, we [had to make](https://questdb.io/blog/making-our-website-source-private/) our website's repo private recently and we haven't moved the docs to a separate repo yet, so it wouldn't be possible to contribute a doc change. But no prob with that as someone from our team will document the function soon. Thanks for your contribution once again!

**Comment by kevinmingtarja** [comment]
Ah I didn't notice that. Okay then!
